spec/sec-class-definitions-patch.html

```diff
@@ -10,268 +10,398 @@ <h2>Syntax</h2>
+    <emu-note>
+      <p>Static semantic rules that depend upon substructure generally do not look into class bodies except for |ClassElementName|s.</p>
+    </emu-note>
+
+    <emu-grammar>
+      AsyncFunctionDeclaration : `async` `function` BindingIdentifier `(` FormalParameters `)` `{` AsyncFunctionBody `}`
+
+      AsyncFunctionDeclaration : `async` `function` `(` FormalParameters `)` `{` AsyncFunctionBody `}`
+
+      AsyncFunctionExpression : `async` `function` `(` FormalParameters `)` `{` AsyncFunctionBody `}`
+
+      AsyncFunctionExpression : `async` `function` BindingIdentifier `(` FormalParameters `)` `{` AsyncFunctionBody `}`
+    </emu-grammar>
     <emu-alg>
-      1. <ins>Return the TopLevelLexicallyDeclaredNames of |StatementList|.</ins>
+      1. Return *false*.
     </emu-alg>
-  </emu-clause>
+    <emu-note>
+      <p>Static semantic rules that depend upon substructure generally do not look into function definitions.</p>
+    </emu-note>
 
-  <emu-clause id="sec-class-definitions-static-semantics-vardeclarednames">
-    <h1>Static Semantics: VarDeclaredNames</h1>
-    <emu-grammar collapsed><ins>ClassBlockStatementList: [empty]</ins></emu-grammar>
+    <emu-grammar>ArrowFunction : ArrowParameters `=>` ConciseBody</emu-grammar>
     <emu-alg>
-      1. <ins>Return a new empty List.</ins>
+      1. Return *false*.
     </emu-alg>
-    <emu-grammar collapsed><ins>ClassBlockStatementList: StatementList</ins></emu-grammar>
+
+    <emu-grammar>
+      AsyncArrowFunction : `async` AsyncArrowBindingIdentifier `=>` AsyncConciseBody
+    </emu-grammar>
     <emu-alg>
-      1. <ins>Return the TopLevelVarDeclaredNames of |StatementList|.</ins>
+      1. Return *false*.
+    </emu-alg>
+
+    <emu-grammar>
+      AsyncArrowFunction : CoverCallExpressionAndAsyncArrowHead `=>` AsyncConciseBody
+    </emu-grammar>
+    <emu-alg>
+      1. Return *false*.
     </emu-alg>
-  </emu-clause>
 
-  <emu-clause id="sec-class-definitions-static-semantics-varscopeddeclarations">
-    <h1>Static Semantics: VarScopedDeclarations</h1>
-    <emu-grammar collapsed><ins>ClassBlockStatementList: [empty]</ins></emu-grammar>
+    <emu-grammar>
+      UnaryExpression : AwaitExpression
+    </emu-grammar>
     <emu-alg>
-      1. <ins>Return a new empty List.</ins>
+      1. Return *true*.
     </emu-alg>
-    <emu-grammar collapsed><ins>ClassBlockStatementList: StatementList</ins></emu-grammar>
+
+    <emu-grammar>
+      ForInOfStatement : `for` `await` `(` LeftHandSideExpression `of` AssignmentExpression `)` Statement
+
+      ForInOfStatement : `for` `await` `(` `var` ForBinding `of` AssignmentExpression `)` Statement
+
+      ForInOfStatement : `for` `await` `(` ForDeclaration `of` AssignmentExpression `)` Statement
+    </emu-grammar>
     <emu-alg>
-      1. <ins>Return the TopLevelVarScopedDeclarations of |StatementList|.</ins>
+      1. Return *true*.
     </emu-alg>
+    </ins>
```

Comment:
Might need constructor declaration too?

Reply:
And also property declaration?

Reply:
Constructors are just MethodDefinitions with the name `constructor`. I'll look into the property definition concern, as that may also have been missed by the class fields proposal.

Reply:
`Contains` gets around this by shunting to `ComputedPropertyContains` for a _ClassTail_. It looks like this was missed in _ContainsArguments_, which would incorrectly look for `arguments` in a _FieldDefinition_'s initializer. In general this isn't a problem because that would be a static error anyways, but it does mean that _ContainsArguments_ incorrectly and unnecessarily recurs through nested _FieldDefinition_s:

```js
class A {
  static B = class {
    static C = class {
      static D = arguments; //
    }
  }
}
```
In this case, `ContainsArguments` would descend all the way to `D` when checking the static semantics of `A`, `B`, and `C`, when only `ContainsArguments` for `C` really needs to do that descent. Instead, `ContainsArguments` should only check a _ComputedPropertyName_ of the _FieldDefinition_. This seems to be a spec bug for class fields, which I can file shortly in the ecma262 repo.

@littledan, can you clarify for me why _FieldInitializer_ is specified with `?Await`? It seems like it should be `~Await`:

![image](https://user-images.githubusercontent.com/3902892/122454373-2d5fdf80-cf60-11eb-9e88-c5d8189ed4a0.png)

While a class field's _ComputedPropertyName_ could contain `await` in an `Await` context, an initializer can _never_ be async since its wrapped in its own implicit function. Plus this doesn't make sense:

```js
async function f(p) {
  return class C {
    x = await p; // when could this be awaited?
  };
}
let C = await f(Promise.resolve(1));
let obj = new C();
obj.x; // what is this value?
```

Reply:
I've filed https://github.com/tc39/ecma262/issues/2438 for the issue with `ContainsArguments`, and https://github.com/tc39/ecma262/issues/2437 for the issue with _FieldDefinition_.

Reply:
I've refactored `ContainsAwait` into two parts to better match `Contains`, adding a `ComputedPropertyContainsAwait` that mirrors `ComputedPropertyContains`. I've also updated `ContainsAwait` to more accurately mirror `Contains` except in the specific places where the algorithms must differ.